fastjet/internal/delaunay/triangle.go

```diff
@@ -0,0 +1,193 @@
+type Triangle struct {
+	children []*Triangle // children are triangles that lead to the removal of this triangle
+	A, B, C  *Point
+}
+
+// Returns a triangle formed out of the three given points
+func NewTriangle(a, b, c *Point) *Triangle {
+	// make it counter clockwise
+	if clockwise(a, b, c) {
+		b, c = c, b
+	}
+	return &Triangle{
+		A: a,
+		B: b,
+		C: c,
+	}
+}
+
+func clockwise(a *Point, b *Point, c *Point) bool {
+	return (b.Y-a.Y)*(c.X-b.X)-(c.Y-b.Y)*(b.X-a.X) > 0
+}
+
+// inTriangle checks whether the point is in the triangle and whether it is on an edge
+// using barycentric coordinates
+func (t *Triangle) inTriangle(p *Point) (inside, edge bool) {
+	barcen1 := ((t.B.Y-t.C.Y)*(p.X-t.C.X) + (t.C.X-t.B.X)*(p.Y-t.C.Y)) / det(t)
+	barcen2 := ((t.C.Y-t.A.Y)*(p.X-t.C.X) + (t.A.X-t.C.X)*(p.Y-t.C.Y)) / det(t)
+	barcen3 := 1 - barcen1 - barcen2
+
+	// inside triangle
+	if barcen1 > 0 && barcen1 < 1 && barcen2 > 0 && barcen2 < 1 && barcen3 > 0 && barcen3 < 1 {
+		return true, false
+	}
+	// either outside triangle or on edge
+	in := barcen1 < 0 || barcen2 < 0 || barcen3 < 0
+	return !in, !in
+}
+
+func det(t *Triangle) float64 {
+	return (t.B.Y-t.C.Y)*(t.A.X-t.C.X) + (t.C.X-t.B.X)*(t.A.Y-t.C.Y)
+}
+
+// inCircumCircle returns whether the point is inside the circumcircle of the triangle
+func (t *Triangle) inCircumCircle(p *Point) bool {
+	x, y := t.centerOfCircumCircle()
+	// r is the squared radius of the circumcircle
+	r := (x-t.A.X)*(x-t.A.X) + (y-t.A.Y)*(y-t.A.Y)
+	// point is in circumcircle if squared distance to center is less than squared radius
+	return (p.X-x)*(p.X-x)+(p.Y-y)*(p.Y-y) < r
+}
+
+// centerOfCircumCircle returns the center of the triangle's circum circle.
+func (t *Triangle) centerOfCircumCircle() (x, y float64) {
+	m1 := (t.A.X - t.B.X) / (t.B.Y - t.A.Y)
+	m2 := (t.A.X - t.C.X) / (t.C.Y - t.A.Y)
+	b1 := (t.A.Y+t.B.Y)*0.5 - m1*(t.A.X+t.B.X)*0.5
+	b2 := (t.A.Y+t.C.Y)*0.5 - m2*(t.A.X+t.C.X)*0.5
+	// x and y are the coordinates of the center of the circumcircle
+	x = (b2 - b1) / (m1 - m2)
+	y = m1*x + b1
+	if math.IsNaN(x) || math.IsNaN(y) {
+		m1 = (t.B.X - t.A.X) / (t.A.Y - t.B.Y)
+		m2 = (t.B.X - t.C.X) / (t.C.Y - t.B.Y)
+		b1 = (t.B.Y+t.A.Y)*0.5 - m1*(t.B.X+t.A.X)*0.5
+		b2 = (t.B.Y+t.C.Y)*0.5 - m2*(t.B.X+t.C.X)*0.5
+		x = (b2 - b1) / (m1 - m2)
+		y = m1*x + b1
+		if math.IsNaN(x) || math.IsNaN(y) {
+			m1 = (t.C.X - t.A.X) / (t.A.Y - t.C.Y)
+			m2 = (t.C.X - t.B.X) / (t.B.Y - t.C.Y)
+			b1 = (t.C.Y+t.A.Y)*0.5 - m1*(t.C.X+t.A.X)*0.5
+			b2 = (t.C.Y+t.B.Y)*0.5 - m2*(t.C.X+t.B.X)*0.5
+			x = (b2 - b1) / (m1 - m2)
+			y = m1*x + b1
+			if math.IsNaN(x) || math.IsNaN(y) {
+				panic(fmt.Errorf("delaunay: error caluclating the circumcenter of triangle " + t.String("T")))
+			}
+		}
+	}
+	return x, y
+}
+
+// appends and updates nearest neighbor
+func appendT(triangles []*Triangle, elems ...*Triangle) []*Triangle {
```

Comment:
If you make a `type triangles []*Triangle` you get `(*triangles).append`, `(*triangles).remove` and `triangles.finalize` as namespaced methods which improves clarity. Then just change the `children` field to `triangles`.

Is this function used somewhere?